### DISCUSSION TITLE
Safe serialisation options - e.g. BSON/Ion

### DISCUSSION THREAD

**Discussion by miili** [pydantic/pydantic#4097] [Feature Request]
JSON has serious design issues serializing often used data formats, e.g. `int`, `float64` or `datetime`. These issues are addressed by serialisation standards like BSON (https://en.wikipedia.org/wiki/BSON) or Ion (https://en.wikipedia.org/wiki/Ion_(serialization_format)), to just name a couple.

I think that pydantic could profit by leveraging more sophisticated and type-safe serialization formats. What do you think?

**Comment by samuelcolvin** [comment]
Happy to help make it easier to deserialise and serialise more formats externally, but JSON is by far the most popular format, so I think the only format we'll have dedicated support for is JSON.

If we started supporting other formats there would be:
* Msgpack
* Pickle
* Cbor
* Toml
* Yaml
* CSV
* Ini

And many more, that's easy outside the remit of pydantic.

**Comment by rafalkrupinski** [comment]
Seems pydantic's focus is validation, but I think there's a room for data serialization projects for pydantic.

- pickle: isn't it supported out of the box?
- TOML: the author of pytoml discourages of using this format
- CSV: tabular format for tree-like structure? 🤔
- YAML: tricky format. I'd be interested in StrictYAML, which might play nicely with pydantic annotations as the schema.

**Reply by samuelcolvin** [reply]
See https://github.com/pydantic/pydantic-core/issues/79.

Although actually on serialisation in particular, it would be very easy to add other formats which are supported by serde, problem is they would be compile time dependencies, and would increase the size of binaries.

Btw, I wouldn't take the author of the `toml` package too seriously, TOML is pretty widely accepted now, and it's been added to the standard library in 3.11.

**Reply by rafalkrupinski** [reply]
> See [pydantic/pydantic-core#79](https://github.com/pydantic/pydantic-core/issues/79).
> 
> Although actually on serialisation in particular, it would be very easy to add other formats which are supported by serde, problem is they would be compile time dependencies, and would increase the size of binaries.

we're already parsing yaml files via python dicts, so it doesn't matter if it's not in core/rust, it won't be any slower. If was a separate module, at least it'd be reusable.
Also having a couple of parse&dump modules would allow you to estimate the demand for different formats among users and help you decide on adding them to the core.

> Btw, I wouldn't take the author of the `toml` package too seriously, TOML is pretty widely accepted now, and it's been added to the standard library in 3.11.

I take seriously someone who put their time into writing that library, since they have experience with it.
AFAIR the problem brought up with TOML shows with bigger files. The python community decided to keep it for pyproject, since it's supposed to be small.
Again, for small files speed wouldn't matter so much and I don't think anyone use it for data transfer, as opposed to small config files, so again it would make sense to keep it on the python side.

Which brings me to the thought that it might be a good idea to implement a binary formats in the core, for those cases where speed really does matter.

Cheers

**Reply by samuelcolvin** [reply]
> we're already parsing yaml files via python dicts, so it doesn't matter if it's not in core/rust, it won't be any slower. 

Parsing yaml (or any other format) in rust and creating python objects only during validation will be significantly faster than other routes.

How much you or other users care is a different matter although.

Parsing in rust might also allow file line numbers in errors, which would be a massive win, but will probably require custom parsers for each format.

**Reply by rafalkrupinski** [reply]
At this time I don't have any use cases that suffer from slow parsing. But I'm just one user.

I'm not sure where are you standing on this - should other formats be coded in rust because it's faster? Or in python, because it makes the binary smaller?

> Parsing in rust might also allow file line numbers in errors [...] but will probably require custom parsers for each format.

Not different from python. We're just to lazy not to use yaml.safe_load() ;)

**Reply by samuelcolvin** [reply]
I'm arguing that as per https://github.com/pydantic/pydantic-core/issues/79, the best solution would be if we provided a way to build parsers in rust that can use/be used by pydantic-core without going via python objects.

But I agree that in the short term, parsing with stable and well known python libraries, then parsing a dict to pydantic is fine.
